### PR TITLE
Switch to XMake for Windows support

### LINKED ISSUE #26: Add support for Windows
I've been trying to get this to work off and on for two years, but just getting Unicorn and Lua to build is proving much more difficult than I thought. I would _really_ like to get this working if possible.

So far my major problem is building and installing Lua and Unicorn properly. I'm fairly sure the code itself is cross-platform, but we'll see.

### PR DESCRIPTION
**Pull request by dargueta**
Closes #26

@metaworm thanks a lot for getting this to work! I've never heard of XMake until now, so I'm not sure how to get this completely working. If you could make some modifications I'd love to mainline this. Namely:

- [ ] Work with Lua 5.1-5.4 and LuaJIT
- [ ] Work on Linux and OSX
- [ ] Please restore the readme to what it was originally, but please do make any modifications you feel you need for the build instructions.
- [ ] Modify the workflow YAML to build on Windows
- [ ] Create an AUTHORS.txt and put our names
- [ ] Update the changelog to indicate this, and bump the version to 1.3.0

### CODE REVIEW

**Comment by metaworm** [comment]
Most of it is fine, and it will take some time

**Comment by dargueta** [comment]
> Most of it is fine, and it will take some time

Thank you! I'm not in any sort of rush. If there's something I can do to split the work with you though I'd be happy to have a learning opportunity.

File: xmake.lua [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        os.mkdir(buildir)
+
+        for _, name in ipairs(archs) do
+            local cpp_name = buildir..'/'..name..'_const.cpp'
+            if not os.exists(cpp_name) then
+                os.execv('python', {'tools/generate_constants.py', path.join(incdir, 'unicorn', name .. '.h'), cpp_name})
+            end
+            target:add('files', cpp_name)
+        end
+    end)
```
Comment (waruqi) [inline xmake.lua:35]: try this?

```lua
add_rules("mode.debug", "mode.release")

local unicorn_archs = {
    "aarch64", "sparc", "sparc", "riscv64", "arm", "m68k",
    "x86_64", "s390x", "mips64", "sparc64", "ppc", "ppc64",
    "mipsel", "riscv32", "mips", "mips64el"}
option("unicorn-arch", {description = "Select unicorn arch", default = "aarch64", values = unicorn_archs})

add_requires("lua 5.4", {configs = {shared = true}})
add_requires("unicorn", {configs = {arch = get_config("unicorn-arch")}})

rule("generate_constants")
    before_buildcmd(function (target, batchcmds, opt)
        local archs = {"arm", "arm64", "mips", "m68k", "sparc", "x86"}
        local includedir = target:pkg("unicorn"):get("sysincludedirs")[1]
        batchcmds:mkdir(target:autogendir())
        for _, name in ipairs(archs) do
            local sourcefile = path.join(target:autogendir(), name .. "_const.cpp")
            batchcmds:show_progress(opt.progress, "${color.build.object}compiling.constants %s", sourcefile)
            batchcmds:vrunv("python", {'tools/generate_constants.py', path.join(includedir, "unicorn", name .. ".h"), sourcefile})
            local objectfile = target:objectfile(sourcefile)
            batchcmds:compile(sourcefile, objectfile)
            table.insert(target:objectfiles(), objectfile)
            batchcmds:add_depfiles(sourcefile)
        end
    end)
rule_end()

target("unicorn54")
    set_kind("shared")
    add_files("src/*.cpp")
    set_languages("c++17")
    add_includedirs("include")
    add_packages("lua", "unicorn")
    add_rules("generate_constants")
```

and update repo first, `xrepo update-repo`

Comment (metaworm) [reply xmake.lua:35]: 我前天还在考虑怎么用xmake获取到unicorn，看了xmake-repo里没有unicorn，没想到今天xmake作者就来给加上了 😄 

搞起来！

Comment (dargueta) [reply xmake.lua:35]: ^ Translation for English-speakers

> The day before yesterday, I was still thinking about how to use xmake to get unicorn. I saw that there is no unicorn in xmake-repo. I didn't expect the author of xmake to add it today 😄 Let's do it!

<hr>

@waruqi if I understand this correctly the unicorn C library is now on XMake, but this Lua binding is _not_, correct?

Comment (waruqi) [reply xmake.lua:35]: > ^ Translation for English-speakers
> 
> > The day before yesterday, I was still thinking about how to use xmake to get unicorn. I saw that there is no unicorn in xmake-repo. I didn't expect the author of xmake to add it today 😄 Let's do it!
> 
> @waruqi if I understand this correctly the unicorn C library is now on XMake, but this Lua binding is _not_, correct?

right, but you can also submit into xmake-repo

**Comment by dargueta** [comment]
No activity in over a year, closing this PR. A lot has changed so please open a new PR if you want to get this in.

**Pull request closed without merging**